### PR TITLE
Added support for snappy to enable prometheus remote write

### PR DESCRIPTION
**Pull request by JustinMason**
Adding a feature.

This adds `snappy` support for the confighttp compression package.
This provides support for enabling the prometheus remote write receiver which only supports snappy.

https://github.com/open-telemetry/opentelemetry-collector/issues/7632

Added unit test coverage and testing integration with prometheus remote write compatibility.

### CODE REVIEW

**Review by paul-magid-sp — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

File: config/confighttp/compression.go [hunk 1]
```diff
@@ -117,6 +118,25 @@ func httpContentDecompressor(h http.Handler, eh func(w http.ResponseWriter, r *h
...
+
+				reqBuf, err := snappy.Decode(nil, compressed)
+				if err != nil {
+					return nil, err
+				}
+
+				sr := &readCloser{
+					reader: bytes.NewReader(reqBuf),
+					closer: body,
+				}
```
Comment (bogdandrutu) [inline config/confighttp/compression.go:136]: You already read the bytes why not closing the body and make the closer no-op?

Comment (JustinMason) [reply config/confighttp/compression.go]: I address this.
Nice suggestion, no need to create my own struct either.

**Review by atoulme — APPROVED** [review]

File: config/confighttp/compression.go [hunk 2]
```diff
@@ -117,6 +118,27 @@ func httpContentDecompressor(h http.Handler, eh func(w http.ResponseWriter, r *h
...
+
+				reqBuf, err := snappy.Decode(nil, compressed)
+				if err != nil {
+					return nil, err
+				}
+
+				sr := io.NopCloser(bytes.NewReader(reqBuf)) 
+
+				return sr, nil
```
Comment (bogdandrutu) [inline config/confighttp/compression.go:140]:
```suggestion
				return io.NopCloser(bytes.NewReader(reqBuf)), nil
```

File: config/confighttp/compression.go [hunk 3]
```diff
@@ -117,6 +118,27 @@ func httpContentDecompressor(h http.Handler, eh func(w http.ResponseWriter, r *h
...
+
+				compressed, err := io.ReadAll(body)
+				if err != nil {
+					return nil, err
+				}
+				
+				err = body.Close()
+				if err != nil {
+					return nil, err
+				}
```
Comment (bogdandrutu) [inline config/confighttp/compression.go:131]:
```suggestion
				if err = body.Close(); err != nil {
					return nil, err
				}
```

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@JustinMason there are some `go fmt` errors, see https://github.com/open-telemetry/opentelemetry-collector/actions/runs/7359299847/job/20036947971?pr=8983

**Comment by atoulme** [comment]
Please take a look at the failing tests.

**Comment by JustinMason** [comment]
> Please take a look at the failing tests.

I think I got them fixed.
My forked version passed except:

```
The error stack is: Error: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Could not find a repository, try using repo upload token', code='not_found')}
```

I was assuming this was because my forked repo isn't registered, not sure.

**Review by codeboten — APPROVED** [review]

**Pull request merged**